### ISSUE TITLE
FixtureLifeCycle(LifeCycle.InstancePerTestCase) Not working with TestFixtureSource

### ISSUE THREAD

**Issue by xadvfh** [nunit/nunit#3715]
I'm using the new `FixtureLifeCycle(LifeCycle.InstancePerTestCase)` with `TestFixtureSource` and am running into an issue where a new instance is not created.

At first, I thought this might be an issue with WireMock.Net but it works fine without `TestFixtureSource`.  It also works fine when using an `IDisposable` class to simulate setup/teardown per test as @jnm2 has mentioned in the past.

In my tests, I am using [WireMock.Net](https://github.com/WireMock-Net/WireMock.Net) to create a new instance of the mock server for each test.  Before `FixtureLifeCycle`, I was starting the server in the constructor of an `IDisposable` class and stopping the server in the `Dispose` method.

In my tests below I'm just writing out the URL of the mock server to show that they are using the same url:port.  Each test should start a server on a unique port number but both tests are using the same port.

```c#
using NUnit.Framework;
using System;
using System.Collections;
using System.Linq;
using WireMock.Server;

namespace NUnitTestProject1
{
    [TestFixture]
    [Parallelizable(ParallelScope.All)]
    [FixtureLifeCycle(LifeCycle.InstancePerTestCase)]
    [TestFixtureSource(typeof(BrowserData))]
    public class TestBase
    {
        protected WireMockServer Server { get; set; }
        protected string driverId;

        public TestBase(string driver)
        {
            driverId = driver;
        }

        [SetUp]
        public void BeforeTest()
        {
            Server = WireMockServer.Start();
        }


        [TearDown]
        public void AfterTest()
        {
            Server?.Stop();
        }
    }


    public class Tests : TestBase
    {
        public Tests(string driver) : base(driver)
        {
        }

        [Test]
        public void Test1()
        {
            TestContext.Out.WriteLine(Server.Urls.First());
        }

        [Test]
        public void Test2()
        {
            TestContext.Out.WriteLine(Server.Urls.First());
        }
    }


    public class BrowserData : IEnumerable
    {
        private readonly string[] _drivers;
        public BrowserData()
        {
            var driverId = "LocalChrome";
            _drivers = Array.ConvertAll(driverId.Split(",", StringSplitOptions.RemoveEmptyEntries), d => d.Trim());
        }

        public IEnumerator GetEnumerator()
        {
            return _drivers.GetEnumerator();
        }
    }
}
```

**Comment by gleb-osokin** [comment]
I can confirm the bug, here's a refined example test that fails:
````csharp   
    [TestFixtureSource(nameof(FixtureArgs))]
    [FixtureLifeCycle(LifeCycle.InstancePerTestCase)]
    public class LifeCycleWithTestFixtureSourceTest
    {
        private readonly int _initialValue;
        private int _value;

        public LifeCycleWithTestFixtureSourceTest(int num)
        {
            _initialValue = num;
            _value = num;
        }

        public static int[] FixtureArgs() => new[] { 1, 42 };

        [Test]
        public void Test1() => Assert.AreEqual(_initialValue, _value++);

        [Test]
        public void Test2() => Assert.AreEqual(_initialValue, _value++);
    }`

**Comment by xadvfh** [comment]
@avilv since you worked on the original implementation would you mind taking a look at this?

**Comment by rprouse** [comment]
@xadvfh I'm planning on doing a 3.13.1 release soon. I'm just tracking a few other bugs and giving it a bit of time for others to come to light.

**Comment by xadvfh** [comment]
sounds great, thank you for your help.

**Comment by rprouse** [comment]
Fixed by #3720

**Comment by Lisaj** [comment]
Hi, i have same problem with TestCaseSource. Is it fixed by this bug to?

**Comment by uchagani** [comment]
It is fixed but we’re waiting on a release which should be by the end of this month. I’m not sure if it has made its way into the nightly build.

**Comment by ivan-menshikov** [comment]
#3720 does not seem to fix the following:

```cs
[FixtureLifeCycle(LifeCycle.InstancePerTestCase)]
[Parallelizable(ParallelScope.All)]
public class FixtureLifeCycleTests
{
	private int _counter = 0;

	[TestCase(0)]
	[TestCase(1)]
	[TestCase(2)]
	[TestCase(3)]
	public void WorksWithTestCaseAttribute(int _)
	{
		Assert.AreEqual(1, Interlocked.Increment(ref _counter));
	}
}
```

**Comment by gleb-osokin** [comment]
@ivan-menshikov this issue relates specifically to TestFixtureSource parametrization.

I have created a new issue #3743, that addresses this new case as well.
